src/main/java/core/stack/MateStack.java

```diff
@@ -1,19 +1,54 @@
 package core.stack;
 
+import java.util.EmptyStackException;
+
 public class MateStack<T> {
-    public void push(T value) {
+    private Node<T> tail;
+    private Node<T> head;
+    private int size;
```

Comment:
You don't need two pointers in a singly-connected list. Remember how you added elements to the head of a singly linked list. Create a node that always points to the top element of the stack. You have to push the elements down.
```suggestion
    private Node<T> top;
    private int size;
```

Reply:
changed